File: RaceControlAndroid/app/src/main/java/com/owlmedia/racecontrol/feature/standings/WdcCalculatorView.kt
package com.owlmedia.racecontrol.feature.standings

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.RowScope
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.EmojiEvents
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Slider
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableFloatStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.owlmedia.racecontrol.R
import com.owlmedia.racecontrol.core.design.Dimens
import com.owlmedia.racecontrol.core.design.RcShapes
import com.owlmedia.racecontrol.core.design.RcTheme
import com.owlmedia.racecontrol.core.design.tabular
import com.owlmedia.racecontrol.core.ui.EmptyState
import com.owlmedia.racecontrol.core.ui.LoadableContent
import com.owlmedia.racecontrol.core.ui.TeamLogo
import com.owlmedia.racecontrol.data.remote.dto.RaceEventDto
import com.owlmedia.racecontrol.data.remote.dto.WdcDriverDto
import com.owlmedia.racecontrol.feature.AppState
import kotlin.math.ceil
import kotlin.math.roundToInt

/**
 * Who can still mathematically win the drivers' championship.
 *
 * Mirrors the web/iOS "WDC calculator": for every driver, the ceiling is
 * their current points plus every remaining point on offer (race + sprint),
 * assuming they win everything left and the leader scores nothing else. That
 * is a best case, not a forecast. The caption under the header exists
 * because a user found the raw numbers confusing without it.
 */
@Composable
fun WdcCalculatorView(
    viewModel: StandingsViewModel,
    appState: AppState,
    onOpenDriver: (String) -> Unit = {},
) {
    val state by viewModel.wdc.collectAsStateWithLifecycle()
    val throughRound by viewModel.wdcThroughRound.collectAsStateWithLifecycle()
    val completedRounds by viewModel.wdcCompletedRounds.collectAsStateWithLifecycle()

    LoadableContent(
        state = state,
        onRetry = { viewModel.load(appState.selectedYear, StandingsMode.WDC, force = true) },
    ) { data ->
        if (data.drivers.isEmpty()) {
            EmptyState(
                icon = Icons.Filled.EmojiEvents,
                title = stringResource(R.string.no_standings_title),
                message = stringResource(R.string.no_standings_message, appState.selectedYear.toString()),
            )
            return@LoadableContent
        }

        val canWinCount = data.drivers.count { it.canWin }

        LazyColumn(
            contentPadding = PaddingValues(Dimens.MD),
            verticalArrangement = Arrangement.spacedBy(Dimens.SM),
        ) {
            if (completedRounds.isNotEmpty()) {
                item(key = "wdc-time-machine") {
                    WdcTimeMachine(
                        completedRounds = completedRounds,
                        throughRound = throughRound,
                        onThroughRoundChange = viewModel::setWdcThroughRound,
                    )
                }
            }
            item(key = "wdc-header") {
                WdcHeader(
                    decided = data.decided,
                    roundsRemaining = data.roundsRemaining,
                    maxRemainingPoints = data.maxRemainingPoints,
                    canWinCount = canWinCount,
                    notes = data.notes,
                )
            }
            item(key = "title-scenarios") {
                TitleScenarioMatrix(appState.selectedYear, throughRound)
            }
            items(
                items = data.drivers,
                key = { it.driverId ?: it.fullName },
                contentType = { "wdc-driver" },
            ) { driver ->
                WdcDriverRow(driver = driver, onOpenDriver = onOpenDriver)
            }
        }
    }
}

/**
 * "Time machine" round scrubber. Once a season is over, the live calculator degenerates to
 * "only the leader can win" for every round, which isn't interesting: this lets a user scrub
 * back to any completed round and see the calculator as it stood then, using that round's
 * actual cumulative points rather than final-season standings.
 *
 * The slider's valid positions are 1..lastCompletedRound, plus one extra stop past the end
 * that represents live standings (throughRound = null). [onThroughRoundChange] only fires when
 * the drag ends, not on every intermediate value, so scrubbing does not spam the network.
 */
@Composable
private fun WdcTimeMachine(
    completedRounds: List<RaceEventDto>,
    throughRound: Int?,
    onThroughRoundChange: (Int?) -> Unit,
) {
    val lastRound = completedRounds.maxOf { it.round }
    val liveStop = (lastRound + 1).toFloat()
    val committedValue = throughRound?.toFloat() ?: liveStop

    var sliderValue by remember(committedValue) { mutableFloatStateOf(committedValue) }
    val displayedRound = sliderValue.roundToInt()
    val isLive = displayedRound > lastRound

    val label = if (isLive) {
        "Viewing: live standings"
    } else {
        val event = completedRounds.find { it.round == displayedRound }
        "Viewing: as of Round $displayedRound (${event?.displayName ?: "Round $displayedRound"})"
    }

    Column(
        modifier = Modifier
            .fillMaxWidth()
            .clip(RcShapes.Medium)
            .background(MaterialTheme.colorScheme.surface)
            .padding(Dimens.SM),
    ) {
        Row(
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.SpaceBetween,
            modifier = Modifier.fillMaxWidth(),
        ) {
            Text(
                text = label,
                style = MaterialTheme.typography.bodyMedium,
                fontWeight = FontWeight.SemiBold,
                color = RcTheme.colors.textPrimary,
                modifier = Modifier.weight(1f),
            )
            if (!isLive) {
                TextButton(onClick = {
                    sliderValue = liveStop
                    onThroughRoundChange(null)
                }) {
                    Text("Jump to live")
                }
            }
        }
        Slider(
            value = sliderValue,
            onValueChange = { sliderValue = it },
            onValueChangeFinished = {
                val round = sliderValue.roundToInt()
                onThroughRoundChange(if (round > lastRound) null else round)
            },
            valueRange = 1f..liveStop,
            steps = (liveStop.toInt() - 1) - 1,
            modifier = Modifier
                .fillMaxWidth()
                .semantics {
                    contentDescription = if (isLive) {
                        "Round scrubber, live standings"
                    } else {
                        "Round scrubber, round $displayedRound of $lastRound"
                    }
                },
        )
        WdcRoundTicks(
            completedRounds = completedRounds,
            displayedRound = displayedRound,
            isLive = isLive,
            onSelectRound = { round ->
                sliderValue = round.toFloat()
                onThroughRoundChange(round)
            },
            onSelectLive = {
                sliderValue = liveStop
                onThroughRoundChange(null)
            },
        )
    }
}

/**
 * Per-round tick marks below the slider. With only a handful of rounds run so far, the
 * slider's few steps are spread across the whole track, so a small drag jumps a
 * disproportionately large visual gap between adjacent rounds. Showing every round's number
 * makes that discreteness explicit (it's a timeline of specific races, not a continuous
 * scale) and gives a precise tap target instead of trying to land a drag exactly on it.
 * Labels thin out once there are too many rounds to print legibly, but the current round and
 * both ends stay labelled. Includes one extra tick for "Live", mirroring the slider's own
 * extra stop past the last completed round.
 */
@Composable
private fun WdcRoundTicks(
    completedRounds: List<RaceEventDto>,
    displayedRound: Int,
    isLive: Boolean,
    onSelectRound: (Int) -> Unit,
    onSelectLive: () -> Unit,
) {
    val labelStride = if (completedRounds.size > 15) {
        ceil(completedRounds.size / 12.0).toInt()
    } else {
        1
    }

    Row(modifier = Modifier.fillMaxWidth()) {
        completedRounds.forEachIndexed { index, event ->
            val isCurrent = !isLive && event.round == displayedRound
            val isPast = event.round < displayedRound
            val showLabel = isCurrent || index == 0 || index == completedRounds.lastIndex || index % labelStride == 0
            WdcRoundTick(
                label = event.round.toString(),
                isCurrent = isCurrent,
                isPast = isPast,
                showLabel = showLabel,
                contentDescription = "Round ${event.round}, ${event.displayName}",
                onClick = { onSelectRound(event.round) },
            )
        }
        WdcRoundTick(
            label = "Live",
            isCurrent = isLive,
            isPast = false,
            showLabel = true,
            contentDescription = "Live standings",
            onClick = onSelectLive,
        )
    }
}

@Composable
private fun RowScope.WdcRoundTick(
    label: String,
    isCurrent: Boolean,
    isPast: Boolean,
    showLabel: Boolean,
    contentDescription: String,
    onClick: () -> Unit,
) {
    Column(
        horizontalAlignment = Alignment.CenterHorizontally,
        modifier = Modifier
            .weight(1f)
            .clickable(onClickLabel = contentDescription, onClick = onClick)
            .padding(vertical = 4.dp),
    ) {
        Box(
            modifier = Modifier
                .size(if (isCurrent) 8.dp else 5.dp)
                .clip(CircleShape)
                .background(
                    when {
                        isCurrent -> RcTheme.colors.racingRed
                        isPast -> RcTheme.colors.racingRed.copy(alpha = 0.5f)
                        else -> RcTheme.colors.surfaceElevated
                    },
                ),
        )
        Spacer(Modifier.height(2.dp))
        Text(
            text = if (showLabel) label else "",
            style = MaterialTheme.typography.labelSmall,
            fontWeight = if (isCurrent) FontWeight.Bold else FontWeight.Normal,
            color = if (isCurrent) RcTheme.colors.racingRed else RcTheme.colors.textTertiary,
        )
    }
}

@Composable
private fun WdcHeader(
    decided: Boolean,
    roundsRemaining: Int,
    maxRemainingPoints: Int,
    canWinCount: Int,
    notes: List<String> = emptyList(),
) {
    val statusText = when {
        decided && roundsRemaining == 0 -> "Season complete: the title is decided."
        decided -> "Mathematically settled: only the leader can still win."
        else -> {
            val roundsWord = if (roundsRemaining == 1) "round" else "rounds"
            val driversWord = if (canWinCount == 1) "driver" else "drivers"
            "$roundsRemaining $roundsWord left · up to $maxRemainingPoints points still on offer · " +
                "$canWinCount $driversWord can still win"
        }
    }

    Column(modifier = Modifier.padding(bottom = Dimens.XS)) {
        Text(
            text = statusText,
            style = MaterialTheme.typography.bodyMedium,
            fontWeight = FontWeight.SemiBold,
            color = RcTheme.colors.textPrimary,
        )
        Spacer(Modifier.height(2.dp))
        Text(
            text = "\"Can win\" is the best-case ceiling (winning every remaining session while the " +
                "leader scores nothing else), not a realistic forecast.",
            style = MaterialTheme.typography.bodySmall,
            color = RcTheme.colors.textTertiary,
        )
        notes.forEach { note ->
            Text(
                text = "• $note",
                style = MaterialTheme.typography.bodySmall,
                color = RcTheme.colors.textTertiary,
            )
        }
    }
}

@Composable
private fun WdcDriverRow(
    driver: WdcDriverDto,
    onOpenDriver: (String) -> Unit,
) {
    val behindLabel = if (driver.pointsBehindLeader == 0.0) "Leader" else "-${formatPoints(driver.pointsBehindLeader)}"

    Column(
        modifier = Modifier
            .fillMaxWidth()
            .clip(RcShapes.Medium)
            .background(MaterialTheme.colorScheme.surface)
            .padding(Dimens.SM)
            .semantics(mergeDescendants = true) {
                contentDescription = buildString {
                    append("Position ${driver.position ?: 0}, ${driver.fullName}")
                    if (!driver.teamName.isNullOrBlank()) append(", ${driver.teamName}")
                    append(", ${formatPoints(driver.points)} points")
                    append(", ${if (driver.pointsBehindLeader == 0.0) "leader" else "${formatPoints(driver.pointsBehindLeader)} behind leader"}")
                    append(", max possible ${formatPoints(driver.maxPoints)} points")
                    append(if (driver.canWin) ", can still win" else ", cannot win")
                }
            },
    ) {
        Row(verticalAlignment = Alignment.CenterVertically) {
            Text(
                text = (driver.position ?: 0).toString(),
                style = MaterialTheme.typography.titleMedium.tabular(),
                fontWeight = FontWeight.Bold,
                color = if ((driver.position ?: 0) <= 3) RcTheme.colors.gold else RcTheme.colors.textSecondary,
                modifier = Modifier.width(28.dp),
            )
            TeamLogo(url = driver.teamLogoUrl, modifier = Modifier.padding(end = Dimens.SM))
            Column(Modifier.weight(1f)) {
                Text(
                    text = driver.fullName,
                    style = MaterialTheme.typography.bodyLarge,
                    fontWeight = FontWeight.SemiBold,
                    color = if (driver.driverId != null) RcTheme.colors.info else RcTheme.colors.textPrimary,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                    modifier = Modifier.clickable(enabled = driver.driverId != null) {
                        onOpenDriver(driver.driverId ?: return@clickable)
                    },
                )
                if (!driver.teamName.isNullOrBlank()) {
                    Text(
                        text = driver.teamName,
                        style = MaterialTheme.typography.bodySmall,
                        color = RcTheme.colors.textSecondary,
                        maxLines = 1,
                        overflow = TextOverflow.Ellipsis,
                    )
                }
            }
            CanWinPill(canWin = driver.canWin)
        }
        Spacer(Modifier.height(6.dp))
        Row(horizontalArrangement = Arrangement.spacedBy(Dimens.MD)) {
            WdcStat(label = "Points", value = formatPoints(driver.points))
            WdcStat(label = "Behind", value = behindLabel)
            WdcStat(label = "Max possible", value = formatPoints(driver.maxPoints))
        }
    }
}

@Composable
private fun WdcStat(label: String, value: String) {
    Column {
        Text(
            text = label,
            style = MaterialTheme.typography.labelSmall,
            color = RcTheme.colors.textTertiary,
        )
        Text(
            text = value,
            style = MaterialTheme.typography.labelLarge.tabular(),
            fontWeight = FontWeight.Bold,
            color = RcTheme.colors.textPrimary,
        )
    }
}

@Composable
private fun CanWinPill(canWin: Boolean) {
    val background = if (canWin) RcTheme.colors.positive.copy(alpha = 0.2f) else RcTheme.colors.surfaceElevated
    val textColor = if (canWin) RcTheme.colors.positive else RcTheme.colors.textTertiary
    val label = if (canWin) "Can win" else "Can't win"

    Box(
        modifier = Modifier
            .clip(RcShapes.Small)
            .background(background)
            .padding(horizontal = Dimens.SM, vertical = 4.dp),
    ) {
        Text(
            text = label,
            style = MaterialTheme.typography.labelSmall,
            fontWeight = FontWeight.Bold,
            color = textColor,
        )
    }
}

/** Points without a trailing ".0": F1 points are integers apart from half-points races. */
private fun formatPoints(value: Double): String =
    if (value % 1.0 == 0.0) value.toInt().toString() else value.toString()

File: RaceControlApp/RaceControl/Features/Standings/WdcCalculatorView.swift
import SwiftUI

/// Drivers' championship "who can still win" calculator: for each driver still
/// in contention, shows their mathematical ceiling if they won every remaining
/// point-scoring opportunity while the current leader scored nothing else.
struct WdcCalculatorView: View {
    let year: Int
    @StateObject private var vm = WdcCalculatorViewModel()
    /// Local slider position, decoupled from the network-backed
    /// `vm.selectedRound` so dragging doesn't trigger a reload on every tick.
    @State private var scrubValue: Double = 1

    var body: some View {
        LoadableView(state: vm.state) {
            await vm.load(year: year)
        } content: { data in
            content(data)
        }
        .background(Theme.Palette.background)
        .task(id: year) { await vm.load(year: year) }
    }

    private func content(_ data: WdcCalculator) -> some View {
        ScrollView {
            VStack(spacing: Theme.Space.sm) {
                timeMachineControl(data)
                statusHeader(data)
                pointsBreakdown
                TitleScenarioMatrixView(year: year, throughRound: vm.selectedRound)
                LazyVStack(spacing: Theme.Space.sm) {
                    ForEach(data.drivers) { driver in
                        WdcDriverRow(driver: driver, year: year, driversById: vm.driversById)
                    }
                }
                notesFootnote(data)
            }
            .padding(Theme.Space.md)
        }
    }

    /// Known scoring approximations the backend flags for this calculation
    /// (e.g. no countback tie-break, flat modern sprint-points assumption
    /// applied to historical seasons). Backend-provided and optional, so
    /// nothing renders until/unless the server sends notes.
    @ViewBuilder
    private func notesFootnote(_ data: WdcCalculator) -> some View {
        if let notes = data.notes, !notes.isEmpty {
            VStack(alignment: .leading, spacing: 2) {
                ForEach(notes, id: \.self) { note in
                    Text(note)
                }
            }
            .font(.caption2)
            .foregroundStyle(Theme.Palette.textSecondary)
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding(.top, Theme.Space.xs)
        }
    }

    @ViewBuilder
    private func timeMachineControl(_ data: WdcCalculator) -> some View {
        let rounds = vm.completedRounds
        if let lastRound = rounds.map(\.round).max() {
            VStack(alignment: .leading, spacing: Theme.Space.xs) {
                HStack {
                    Text("TIME MACHINE")
                        .font(.caption.weight(.bold)).tracking(1)
                        .foregroundStyle(Theme.Palette.textSecondary)
                    Spacer()
                    if vm.selectedRound != nil {
                        Button("Jump to Live") {
                            Task { await vm.load(year: year, through: nil) }
                        }
                        .font(.caption.weight(.semibold))
                        .foregroundStyle(Theme.Palette.racingRed)
                    }
                }
                Text(viewingLabel(lastRound: lastRound))
                    .font(.subheadline.weight(.semibold))
                    .foregroundStyle(Theme.Palette.textPrimary)
                    .lineLimit(1)
                Slider(
                    value: $scrubValue,
                    in: 1...Double(lastRound),
                    step: 1,
                    onEditingChanged: { editing in
                        guard !editing else { return }
                        let round = Int(scrubValue.rounded())
                        Task { await vm.load(year: year, through: round) }
                    }
                )
                .tint(Theme.Palette.racingRed)
                timelineTicks(rounds: rounds)
            }
            .padding(Theme.Space.md)
            .background(Theme.Palette.surface, in: RoundedRectangle(cornerRadius: Theme.Radius.md))
            .overlay(RoundedRectangle(cornerRadius: Theme.Radius.md).stroke(Theme.Palette.stroke, lineWidth: 1))
            .onAppear {
                scrubValue = Double(vm.selectedRound ?? lastRound)
            }
        }
    }

    /// Per-round markers below the slider: with only a handful of rounds run
    /// so far, the slider's few steps are spread across the whole track, so
    /// a small drag jumps a disproportionately large visual gap between
    /// adjacent rounds. Showing every round's number makes that discreteness
    /// explicit (it's a timeline of specific races, not a continuous scale)
    /// and gives a precise tap target instead of trying to land a drag
    /// exactly on it. Labels thin out once there are too many rounds to
    /// print legibly, but the current round and both ends stay labelled.
    @ViewBuilder
    private func timelineTicks(rounds: [RaceEvent]) -> some View {
        let labelStride = rounds.count > 15 ? Int((Double(rounds.count) / 12.0).rounded(.up)) : 1
        let currentRound = Int(scrubValue.rounded())
        HStack(spacing: 2) {
            ForEach(Array(rounds.enumerated()), id: \.offset) { index, event in
                let isCurrent = event.round == currentRound
                let isPast = event.round < currentRound
                let showLabel = isCurrent || index == 0 || index == rounds.count - 1 || index % labelStride == 0
                Button {
                    scrubValue = Double(event.round)
                    Task { await vm.load(year: year, through: event.round) }
                } label: {
                    VStack(spacing: 2) {
                        Circle()
                            .fill(isCurrent ? Theme.Palette.racingRed : (isPast ? Theme.Palette.racingRed.opacity(0.5) : Theme.Palette.stroke))
                            .frame(width: isCurrent ? 8 : 5, height: isCurrent ? 8 : 5)
                        Text(showLabel ? "\(event.round)" : "")
                            .font(.system(size: 9, weight: isCurrent ? .semibold : .regular))
                            .foregroundStyle(isCurrent ? Theme.Palette.racingRed : Theme.Palette.textTertiary)
                            .frame(height: 10)
                    }
                    .frame(maxWidth: .infinity)
                }
                .buttonStyle(.plain)
                .accessibilityLabel("Round \(event.round), \(event.displayName)")
            }
        }
    }

    private func viewingLabel(lastRound: Int) -> String {
        guard let round = vm.selectedRound else {
            return "Viewing: Live standings"
        }
        if let raceName = vm.completedRounds.first(where: { $0.round == round })?.displayName {
            return "Viewing: as of Round \(round) (\(raceName))"
        }
        return "Viewing: as of Round \(round)"
    }

    @ViewBuilder
    private func statusHeader(_ data: WdcCalculator) -> some View {
        let canWinCount = data.drivers.filter(\.canWin).count
        VStack(alignment: .leading, spacing: 4) {
            Text(statusLine(data, canWinCount: canWinCount))
                .font(.subheadline.weight(.semibold))
                .foregroundStyle(Theme.Palette.textPrimary)
            Text("\u{201C}Can win\u{201D} is the best-case ceiling (winning every remaining session while the leader scores nothing else), not a realistic forecast.")
                .font(.caption)
                .foregroundStyle(Theme.Palette.textSecondary)
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(Theme.Space.md)
        .background(Theme.Palette.surface, in: RoundedRectangle(cornerRadius: Theme.Radius.md))
        .overlay(RoundedRectangle(cornerRadius: Theme.Radius.md).stroke(Theme.Palette.stroke, lineWidth: 1))
    }

    private func statusLine(_ data: WdcCalculator, canWinCount: Int) -> String {
        if data.decided && data.roundsRemaining == 0 {
            return "Season complete: the title is decided."
        }
        if data.decided {
            return "Mathematically settled: only the leader can still win."
        }
        let roundWord = data.roundsRemaining == 1 ? "round" : "rounds"
        let driverWord = canWinCount == 1 ? "driver" : "drivers"
        return "\(data.roundsRemaining) \(roundWord) left · up to \(data.maxRemainingPoints) points still on offer · \(canWinCount) \(driverWord) can still win"
    }

    private var pointsBreakdown: some View {
        DisclosureGroup {
            VStack(alignment: .leading, spacing: Theme.Space.xs) {
                Text("Race (P1–P10): 25, 18, 15, 12, 10, 8, 6, 4, 2, 1")
                Text("Sprint (P1–P8): 8, 7, 6, 5, 4, 3, 2, 1")
                Text("Fastest lap bonus: +1 (top 10 finishers, race only)")
            }
            .font(.caption)
            .foregroundStyle(Theme.Palette.textSecondary)
            .padding(.top, Theme.Space.sm)
            .frame(maxWidth: .infinity, alignment: .leading)
        } label: {
            Text("POINTS BREAKDOWN")
                .font(.caption.weight(.bold)).tracking(1)
                .foregroundStyle(Theme.Palette.textSecondary)
        }
        .padding(Theme.Space.md)
        .background(Theme.Palette.surface, in: RoundedRectangle(cornerRadius: Theme.Radius.md))
        .overlay(RoundedRectangle(cornerRadius: Theme.Radius.md).stroke(Theme.Palette.stroke, lineWidth: 1))
    }
}

private struct WdcDriverRow: View {
    let driver: WdcDriverEntry
    let year: Int
    let driversById: [String: Driver]

    private var matchedDriver: Driver? {
        guard let id = driver.driverId else { return nil }
        return driversById[id]
    }

    var body: some View {
        HStack(spacing: Theme.Space.md) {
            PositionBadge(text: String(driver.position ?? 0), highlight: (driver.position ?? 0) <= 3)
            TeamAccentBar(color: .team(driver.teamColor))
            TeamLogoView(url: driver.teamLogoUrl, size: 28)

            VStack(alignment: .leading, spacing: 2) {
                nameLabel
                Text(driver.teamName ?? "")
                    .font(.caption)
                    .foregroundStyle(Theme.Palette.textSecondary)
                    .lineLimit(1)
            }

            Spacer()

            VStack(alignment: .trailing, spacing: 2) {
                HStack(alignment: .firstTextBaseline, spacing: Theme.Space.sm) {
                    VStack(alignment: .trailing, spacing: 0) {
                        Text(numberLabel(driver.points))
                            .font(.system(.subheadline, design: .rounded).weight(.bold))
                            .monospacedDigit()
                            .foregroundStyle(Theme.Palette.textPrimary)
                        Text("PTS").font(.caption2).foregroundStyle(Theme.Palette.textTertiary)
                    }
                    VStack(alignment: .trailing, spacing: 0) {
                        Text(behindLabel)
                            .font(.caption.weight(.semibold))
                            .foregroundStyle(driver.pointsBehindLeader == 0 ? Theme.Palette.positive : Theme.Palette.textSecondary)
                        Text("MAX \(numberLabel(driver.maxPoints))")
                            .font(.caption2)
                            .foregroundStyle(Theme.Palette.textTertiary)
                    }
                }
                canWinPill
            }
        }
        .padding(Theme.Space.md)
        .background(Theme.Palette.surface, in: RoundedRectangle(cornerRadius: Theme.Radius.md))
        .overlay(RoundedRectangle(cornerRadius: Theme.Radius.md).stroke(Theme.Palette.stroke, lineWidth: 1))
    }

    @ViewBuilder
    private var nameLabel: some View {
        if let matchedDriver {
            NavigationLink {
                DriverDetailView(year: year, driver: matchedDriver)
            } label: {
                Text(driver.fullName)
                    .font(.headline)
                    .foregroundStyle(Theme.Palette.textPrimary)
                    .lineLimit(1)
            }
        } else {
            Text(driver.fullName)
                .font(.headline)
                .foregroundStyle(Theme.Palette.textPrimary)
                .lineLimit(1)
        }
    }

    private var behindLabel: String {
        driver.pointsBehindLeader == 0 ? "Leader" : "-\(numberLabel(driver.pointsBehindLeader))"
    }

    private var canWinPill: some View {
        Text(driver.canWin ? "Can win" : "Can't win")
            .font(.caption2.weight(.bold))
            .foregroundStyle(driver.canWin ? Theme.Palette.positive : Theme.Palette.textTertiary)
            .padding(.horizontal, 8)
            .padding(.vertical, 3)
            .background(
                (driver.canWin ? Theme.Palette.positive : Theme.Palette.textTertiary).opacity(0.15),
                in: Capsule()
            )
    }

    private func numberLabel(_ value: Double) -> String {
        value.truncatingRemainder(dividingBy: 1) == 0 ? String(Int(value)) : String(format: "%.1f", value)
    }
}

@MainActor
final class WdcCalculatorViewModel: ObservableObject {
    @Published var state: Loadable<WdcCalculator> = .idle
    @Published var driversById: [String: Driver] = [:]
    /// Completed events for the season, sorted by round, used to build the
    /// time-machine scrubber's valid range and race-name labels.
    @Published var completedRounds: [RaceEvent] = []
    /// nil = live standings. Non-nil = viewing the calculator as it stood at
    /// the end of that round.
    @Published var selectedRound: Int? = nil
    private var loadedYear: Int?
    /// Wrapped so we can distinguish "never loaded" (nil) from "loaded for
    /// live standings" (Optional(nil)).
    private var loadedRound: Int?? = nil
    /// Tracks which year we've already fired the background cache warm-up
    /// for, so it only happens once per year rather than on every reload.
    private var warmedYear: Int?

    func load(year: Int) async {
        await load(year: year, through: selectedRound)
    }

    func load(year: Int, through round: Int?) async {
        if loadedYear == year, loadedRound == Optional(round), case .loaded = state { return }
        state = .loading
        selectedRound = round
        do {
            state = .loaded(try await APIClient.shared.wdcCalculator(year: year, throughRound: round))
            loadedYear = year
            loadedRound = round
        } catch {
            state = .failed((error as? APIError)?.errorDescription ?? error.localizedDescription)
        }
        // Secondary lookups for driver-detail navigation and the round
        // scrubber; failure shouldn't block the main screen.
        if let drivers = try? await APIClient.shared.drivers(year: year) {
            driversById = Dictionary(uniqueKeysWithValues: drivers.map { ($0.driverId, $0) })
        }
        if let schedule = try? await APIClient.shared.schedule(year: year) {
            completedRounds = schedule.filter(\.completed).sorted { $0.round < $1.round }
        }
        warmTimeMachineCache(year: year)
    }

    /// Fires a background, best-effort request for a historical round as
    /// soon as we know the season has one, so the backend's shared
    /// per-season computation happens during page load instead of during
    /// the user's first slider drag. The backend caches that computation per
    /// year regardless of which round is requested, so warming any single
    /// round makes the *entire* scrubber fast, not just that one round. This
    /// result is discarded, it's purely to warm the server-side cache.
    private func warmTimeMachineCache(year: Int) {
        guard warmedYear != year, let lastRound = completedRounds.last?.round else { return }
        warmedYear = year
        Task.detached {
            _ = try? await APIClient.shared.wdcCalculator(year: year, throughRound: lastRound)
        }
    }
}
